CIP-2551/CIP-2551.md

```diff
@@ -0,0 +1,271 @@
+
+The Ed25519 curve is the elliptic curve defined by the equation:
+
+```
+-x^2 + y^2 = 1 - (121665/121666) * x^2 * y^2
+```
+
+This is a twisted Edwards curve.
+
+Field elements are serialized in big endian notation, taking up 32
+bytes each.  This leaves one unused bit, which should be set to zero.
+
+Group elements can be serialized in compressed form (just the
+x-coordinate), or in uncompressed form (both the x-coordinate and
+the y-coordinate).  A group element encoding consists of 65 bytes in
+compressed form, or 129 bytes in uncompressed form. The first byte should
+have all bits except the three least significant bits set to zero. Let
+b0 denote the least significant bit, b1 the second least significant,
+and b2 the third least significant.
+
+ * b0 is 1 if and only if the group element is in compressed form. 
+ * b1 is 1 if and only if the element is the point at infinity.
+   (The rest of the encoding after b1 should be zero in case b1 1 is 1.)
+ * b2 is 1 if and only if the element is in compressed form and not the
+   point at infinity and the point is "positive," meaning that its y
+   coordinate is lexicographically greater than the y coordinate of the
+   other point on the curve with the same x coordinate.
+
+A compressed group element encoding consists of the one byte
+containing the three flags, followed by field element encoding of the x
+coordinate. An uncompressed group element encoding consists of the one
+byte containing the three flags, followed by two field element encodings,
+first the x coordinate and then the y coordinate.
+
+## Rationale
+
+The inclusion of `Ed25519_scale_and_add` is to allow for an efficient
+implementation of taking linear combinations of group elements. This
+is important for efficient proof verification for Halo 2's inner
+product argument. For example, this operation can be implemented using
+[Pippenger's algorithm](https://jbootle.github.io/Misc/pippenger.pdf).
+Likely, an efficient implementation of this operation would be useful
+for other applications of this curve, too.
+
+
+## Backward Compatibility
+
+There are no expected backward compatibility issues.
+
+## Test Cases
+
+Property test the group, field, and vector space axioms. Also property
+test that `Ed25519_scale_and_add` produces the same result as zipping
+the lists with `Ed25519_scale` and then folding them with `Ed25518_G_add`.
+
+Provide golden tests for serialization and deserialization. These are unit
+tests which provide explicit inputs to serialization and deserialization
+and explicit expectations as to the results. Also provide round trip
+property tests for serialization and deserialization.
+
+Provide round trip property tests for `Ed25519_F_to_integer` and
+`Ed25519_F_from_integer`.  The round trip properties for these functions
+are that
+
+```
+Ed25519_F_from_integer . Ed25519_F_to_integer = id,
+```
+
+and that for all integers `x`,
+
+```
+Ed25519_F_to_integer (Ed25519_F_from_integer x) ≡ x mod 2^255 - 19.
```

Comment:
So from integer computes the modular reduction? If that is the case, it should be mentioned.